analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/ArithmeticCalculator.cs

```diff
@@ -74,6 +74,26 @@ internal static class ArithmeticCalculator
         }
     }
 
+    public static NumberConstraint PropertyReferenceConstraint(ProgramState state, IPropertyReferenceOperationWrapper operation, ISymbol instanceSymbol)
+    {
+        if (operation.Instance is not null
+            && operation.Property.Name is nameof(Array.Length) or nameof(List<int>.Count))
+        {
+            if (instanceSymbol is not null
+                && state[instanceSymbol]?.Constraint<CollectionConstraint>() is { } constraint)
+            {
+                return constraint == CollectionConstraint.Empty
+                    ? NumberConstraint.From(0)
+                    : NumberConstraint.From(1, null);
+            }
+            else if (operation.Instance.Type.DerivesOrImplementsAny(CollectionTracker.CollectionTypes))
+            {
+                return NumberConstraint.From(0, null);
+            }
+        }
+        return null;
+    }
+
```

Comment:
What's your reasoning to put this in `ArithmeticCalculator`? I would have put it in `CollectionTracker` or `PropertyReference`.

Reply:
I was thinking that this returns a NumberConstraint, so it makes more sense to put it here, even if it reads collection constraints.
What do you think?

Reply:
I'd rather coupe it with collections. It's not doing arithmetics nor is it calculating something.